### PR TITLE
feat(policy): Add CTP support for TCP/TLS listeners

### LINKED ISSUE #3163: Support ClientTrafficPolicy for TCP Listener
*Description*:
> TCP Listener can't be configured with `EnableProxyProtocol` / `IdleTimeout` / `ConnectionLimit` / `PerConnectionBufferLimit`, but those fields are already available in ClientTrafficPolicy.

If ClientTrafficPolicy can be attached to TCP Listener, less EnvoyPatchPolicy will be used to extend EG.

### PR DESCRIPTION
**Pull request by aoledk**
**What this PR does / why we need it**:

Enable user to attach ClientTrafficPolicy [^1] to TCP/TLS listeners, available fields are:

- tcpKeepalive
- connection
- enableProxyProtocol
- tls
- ~~timeout~~ (new field will be needed to specify tcp idle timeout in clientTimeout [^2], raised in other issue #3343)

**Which issue(s) this PR fixes**:

Fixes #3163

[^1]: https://gateway.envoyproxy.io/v1.0.1/api/extension_types/#clienttrafficpolicyspec
[^2]: https://gateway.envoyproxy.io/latest/api/extension_types/#clienttimeout

### CODE REVIEW

File: internal/gatewayapi/clienttrafficpolicy.go [hunk 1]
```diff
@@ -327,6 +324,14 @@ func validatePortOverlapForClientTrafficPolicy(l *ListenerContext, xds *ir.Xds,
 		}
 	}
 
+	var tcpIR *ir.TCPListener
```
Comment (arkodg) [inline internal/gatewayapi/clienttrafficpolicy.go:327]: `tcpIR` is unused here, lets add it when needed ?

**Comment by aoledk** [comment]
@arkodg Do you think whether it's necessary to split this big PR to multiple small PRs for easier review ? Each small PR for single field in CTP.

**Comment by arkodg** [comment]
> @arkodg Do you think whether it's necessary to split this big PR to multiple small PRs for easier review ? Each small PR for single field in CTP.

sorry for the delay in reviewing this one, on it !

File: internal/xds/translator/translator.go [hunk 2]
```diff
@@ -484,6 +484,13 @@ func processTCPListenerXdsTranslation(tCtx *types.ResourceVersionTable, tcpListe
 			}
 		}
 
+		// Add the proxy protocol filter if needed
+		// TODO: should make sure all listeners that will be translated into same xDS listener have
+		// same EnableProxyProtocol value, otherwise listeners with EnableProxyProtocol=false will
+		// never accept connection, because listeners with EnableProxyProtocol=true has configured
+		// proxy protocol listener filter for xDS listener, all connection must have ProxyProtocol header.
+		patchProxyProtocolFilter(xdsListener, tcpListener.EnableProxyProtocol)
```
Comment (arkodg) [inline internal/xds/translator/translator.go:492]: why is this added twice ?

Comment (arkodg) [reply internal/xds/translator/translator.go:492]: ~~oh nvm, this is tcp 🙈 ~~

**Review by arkodg — APPROVED** [review]
LGTM thanks !

**Review by zhaohuabing — APPROVED** [review]
LGTM thanks for adding this!
